1.Copy the code from the lesson to create a dataframe full of student grades.

    a.Create a column named passing_english that indicates whether each student has a passing grade in english.

    b.Sort the english grades by the passing_english column. How are duplicates handled?
    they are ordered true or false then by index

    c.Sort the english grades first by passing_english and then by student name. All the students that are failing english should be first,
    and within the students that are failing english they should be ordered alphabetically. The same should be true for the students passing english.
    (Hint: you can pass a list to the .sort_values method)

    d.Sort the english grades first by passing_english, and then by the actual english grade, similar to how we did in the last step.

    e.Calculate each students overall grade and add it as a column on the dataframe. The overall grade is the average of the math, english, and reading grades.

2.Load the mpg dataset. Read the documentation for the dataset and use it for the following questions:

    a.How many rows and columns are there?

    b.What are the data types of each column?

    c.Summarize the dataframe with .info and .describe

    d.Rename the cty column to city.

    e.Rename the hwy column to highway.

    f.Do any cars have better city mileage than highway mileage?

    g.Create a column named mileage_difference this column should contain the difference between highway and city mileage for each car.

    h.Which car (or cars) has the highest mileage difference?

    i.Which compact class car has the lowest highway mileage? The best?

    j.Create a column named average_mileage that is the mean of the city and highway mileage.

    k.Which dodge car has the best average mileage? The worst?

3.Load the Mammals dataset. Read the documentation for it, and use the data to answer these questions:

    a.How many rows and columns are there?

    b.What are the data types?

    c.Summarize the dataframe with .info and .describe

    d.What is the the weight of the fastest animal?

    e.What is the overal percentage of specials?

    f.How many animals are hoppers that are above the median speed? What percentage is this?

In [248]:
from pydataset import data
import pandas as pd
import numpy as np

np.random.seed(123)

students = ['Sally', 'Jane', 'Suzie', 'Billy', 'Ada', 'John', 'Thomas',
            'Marie', 'Albert', 'Richard', 'Isaac', 'Alan']

# randomly generate scores for each student for each subject
# note that all the values need to have the same length here
math_grades = np.random.randint(low=60, high=100, size=len(students))
english_grades = np.random.randint(low=60, high=100, size=len(students))
reading_grades = np.random.randint(low=60, high=100, size=len(students))

df = pd.DataFrame({'name': students,
                   'math': math_grades,
                   'english': english_grades,
                   'reading': reading_grades})

type(df)


pandas.core.frame.DataFrame

 a.Create a column named passing_english that indicates whether each student has a passing grade in english.

In [249]:
df['passing_english'] = df.english > 70


b.Sort the english grades by the passing_english column. How are duplicates handled?
    they are ordered true or false then by index

In [250]:
df.sort_values(by="passing_english")

name  math  english  reading  passing_english
6    Thomas    82       64       81            False
7     Marie    93       63       90            False
8    Albert    92       62       87            False
11     Alan    92       62       72            False
0     Sally    62       85       80             True
1      Jane    88       79       67             True
2     Suzie    94       74       95             True
3     Billy    98       96       88             True
4       Ada    77       92       98             True
5      John    79       76       93             True
9   Richard    69       80       94             True
10    Isaac    92       99       93             True

c.Sort the english grades first by passing_english and then by student name. All the students that are failing english should be first,
    and within the students that are failing english they should be ordered alphabetically. The same should be true for the students passing english.
    (Hint: you can pass a list to the .sort_values method)

In [251]:
df.sort_values(by='name', ascending=True).sort_values(by="passing_english")


name  math  english  reading  passing_english
11     Alan    92       62       72            False
8    Albert    92       62       87            False
7     Marie    93       63       90            False
6    Thomas    82       64       81            False
4       Ada    77       92       98             True
3     Billy    98       96       88             True
10    Isaac    92       99       93             True
1      Jane    88       79       67             True
5      John    79       76       93             True
9   Richard    69       80       94             True
0     Sally    62       85       80             True
2     Suzie    94       74       95             True

d.Sort the english grades first by passing_english, and then by the actual english grade, similar to how we did in the last step.

In [252]:
df.sort_values(by="passing_english").sort_values(by='english', ascending=False)

name  math  english  reading  passing_english
10    Isaac    92       99       93             True
3     Billy    98       96       88             True
4       Ada    77       92       98             True
0     Sally    62       85       80             True
9   Richard    69       80       94             True
1      Jane    88       79       67             True
5      John    79       76       93             True
2     Suzie    94       74       95             True
6    Thomas    82       64       81            False
7     Marie    93       63       90            False
8    Albert    92       62       87            False
11     Alan    92       62       72            False

e.Calculate each students overall grade and add it as a column on the dataframe. The overall grade is the average of the math, english, and reading grades.

In [253]:
df['overall_grade'] = round((df.english + df.math +df.reading) / 3)

In [254]:
df

name  math  english  reading  passing_english  overall_grade
0     Sally    62       85       80             True           76.0
1      Jane    88       79       67             True           78.0
2     Suzie    94       74       95             True           88.0
3     Billy    98       96       88             True           94.0
4       Ada    77       92       98             True           89.0
5      John    79       76       93             True           83.0
6    Thomas    82       64       81            False           76.0
7     Marie    93       63       90            False           82.0
8    Albert    92       62       87            False           80.0
9   Richard    69       80       94             True           81.0
10    Isaac    92       99       93             True           95.0
11     Alan    92       62       72            False           75.0

In [255]:
# Load the dataset and store it in the variable mpg.

from pydataset import data
mpg_df = data('mpg')
mpg_df



manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

a.How many rows and columns are there?
234 rows and 11 columns

b.What are the data types of each column?

In [256]:
mpg_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 234 entries, 1 to 234
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   manufacturer  234 non-null    object 
 1   model         234 non-null    object 
 2   displ         234 non-null    float64
 3   year          234 non-null    int64  
 4   cyl           234 non-null    int64  
 5   trans         234 non-null    object 
 6   drv           234 non-null    object 
 7   cty           234 non-null    int64  
 8   hwy           234 non-null    int64  
 9   fl            234 non-null    object 
 10  class         234 non-null    object 
dtypes: float64(1), int64(4), object(6)
memory usage: 21.9+ KB


In [257]:
mpg_df

manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

c.Summarize the dataframe with .info and .describe

In [258]:
mpg_df.describe()

displ         year         cyl         cty         hwy
count  234.000000   234.000000  234.000000  234.000000  234.000000
mean     3.471795  2003.500000    5.888889   16.858974   23.440171
std      1.291959     4.509646    1.611534    4.255946    5.954643
min      1.600000  1999.000000    4.000000    9.000000   12.000000
25%      2.400000  1999.000000    4.000000   14.000000   18.000000
50%      3.300000  2003.500000    6.000000   17.000000   24.000000
75%      4.600000  2008.000000    8.000000   19.000000   27.000000
max      7.000000  2008.000000    8.000000   35.000000   44.000000

 d.Rename the cty column to city.

In [259]:
mpg_df = mpg_df.rename(columns={'cty': 'city'})


e.Rename the hwy column to highway.

In [260]:
mpg_df = mpg_df.rename(columns={'hwy': 'highway'})

 f.Do any cars have better city mileage than highway mileage?
 no

In [261]:
mpg_df[mpg_df.city > mpg_df.highway]


Empty DataFrame
Columns: [manufacturer, model, displ, year, cyl, trans, drv, city, highway, fl, class]
Index: []

g.Create a column named mileage_difference this column should contain the difference between highway and city mileage for each car.

In [262]:
mpg_df['mileage_difference'] = abs(mpg_df.highway - mpg_df.city)


h.Which car (or cars) has the highest mileage difference?


In [263]:
mpg_df.sort_values("mileage_difference", ascending= False).head(2)

manufacturer       model  displ  year  cyl     trans drv  city  highway  \
107        honda       civic    1.8  2008    4  auto(l5)   f    24       36   
223   volkswagen  new beetle    1.9  1999    4  auto(l4)   f    29       41   

    fl       class  mileage_difference  
107  c  subcompact                  12  
223  d  subcompact                  12

i.Which compact class car has the lowest highway mileage? The best?

In [264]:
compact_car_mpg_df = mpg_df[mpg_df["class"].str.contains("compact")]
compact_car_mpg_df.sort_values("highway")

manufacturer         model  displ  year  cyl       trans drv  city  \
99          ford       mustang    5.4  2008    8  manual(m6)   r    14   
95          ford       mustang    4.6  1999    8    auto(l4)   r    15   
98          ford       mustang    4.6  2008    8    auto(l5)   r    15   
96          ford       mustang    4.6  1999    8  manual(m5)   r    15   
220   volkswagen         jetta    2.8  1999    6    auto(l4)   f    16   
97          ford       mustang    4.6  2008    8  manual(m5)   r    15   
122      hyundai       tiburon    2.7  2008    6  manual(m5)   f    17   
121      hyundai       tiburon    2.7  2008    6  manual(m6)   f    16   
221   volkswagen         jetta    2.8  1999    6  manual(m5)   f    17   
120      hyundai       tiburon    2.7  2008    6    auto(l4)   f    17   
212   volkswagen           gti    2.8  1999    6  manual(m5)   f    17   
94          ford       mustang    4.0  2008    6    auto(l5)   r    16   
172       subaru   impreza awd    2.5  2008    4  manual(m5)   4    19   
92          ford       mustang    3.8  1999    6    auto(l4)   r    18   
170       subaru   impreza awd    2.5  2008    4    auto(s4)   4    20   
14          audi    a4 quattro    3.1  2008    6    auto(s6)   4    17   
13          audi    a4 quattro    2.8  1999    6  manual(m5)   4    17   
12          audi    a4 quattro    2.8  1999    6    auto(l5)   4    15   
9           audi    a4 quattro    1.8  1999    4    auto(l5)   4    16   
15          audi    a4 quattro    3.1  2008    6  manual(m6)   4    15   
91          ford       mustang    3.8  1999    6  manual(m5)   r    18   
168       subaru   impreza awd    2.5  1999    4  manual(m5)   4    19   
167       subaru   impreza awd    2.2  1999    4  manual(m5)   4    19   
166       subaru   impreza awd    2.2  1999    4    auto(l4)   4    21   
225   volkswagen    new beetle    2.0  1999    4    auto(l4)   f    19   
5           audi            a4    2.8  1999    6    auto(l5)   f    16   
116      hyundai       tiburon    2.0  1999    4    auto(l4)   f    19   
6           audi            a4    2.8  1999    6  manual(m5)   f    18   
215   volkswagen         jetta    2.0  1999    4    auto(l4)   f    19   
192       toyota  camry solara    3.0  1999    6  manual(m5)   f    18   
191       toyota  camry solara    3.0  1999    6    auto(l4)   f    18   
93          ford       mustang    4.0  2008    6  manual(m5)   r    17   
209   volkswagen           gti    2.0  1999    4    auto(l4)   f    19   
169       subaru   impreza awd    2.5  1999    4    auto(l4)   4    19   
8           audi    a4 quattro    1.8  1999    4  manual(m5)   4    18   
11          audi    a4 quattro    2.0  2008    4    auto(s6)   4    19   
143       nissan        altima    2.4  1999    4    auto(l4)   f    19   
187       toyota  camry solara    2.2  1999    4    auto(l4)   f    21   
193       toyota  camry solara    3.3  2008    6    auto(s5)   f    18   
171       subaru   impreza awd    2.5  2008    4    auto(s4)   4    20   
7           audi            a4    3.1  2008    6    auto(av)   f    18   
119      hyundai       tiburon    2.0  2008    4    auto(l4)   f    20   
173       subaru   impreza awd    2.5  2008    4  manual(m5)   4    20   
118      hyundai       tiburon    2.0  2008    4  manual(m5)   f    20   
226   volkswagen    new beetle    2.5  2008    5  manual(m5)   f    20   
10          audi    a4 quattro    2.0  2008    4  manual(m6)   4    20   
208   volkswagen           gti    2.0  1999    4  manual(m5)   f    21   
1           audi            a4    1.8  1999    4    auto(l5)   f    18   
211   volkswagen           gti    2.0  2008    4    auto(s6)   f    22   
214   volkswagen         jetta    2.0  1999    4  manual(m5)   f    21   
216   volkswagen         jetta    2.0  2008    4    auto(s6)   f    22   
218   volkswagen         jetta    2.5  2008    5    auto(s6)   f    21   
219   volkswagen         jetta    2.5  2008    5  manual(m5)   f    21   
224   vo

In [265]:
compact_car_mpg_df.sort_values("highway", ascending= False)

manufacturer         model  displ  year  cyl       trans drv  city  \
213   volkswagen         jetta    1.9  1999    4  manual(m5)   f    33   
222   volkswagen    new beetle    1.9  1999    4  manual(m5)   f    35   
223   volkswagen    new beetle    1.9  1999    4    auto(l4)   f    29   
197       toyota       corolla    1.8  2008    4  manual(m5)   f    28   
107        honda         civic    1.8  2008    4    auto(l5)   f    24   
106        honda         civic    1.8  2008    4    auto(l5)   f    25   
198       toyota       corolla    1.8  2008    4    auto(l4)   f    26   
196       toyota       corolla    1.8  1999    4  manual(m5)   f    26   
105        honda         civic    1.8  2008    4  manual(m5)   f    26   
195       toyota       corolla    1.8  1999    4    auto(l4)   f    24   
100        honda         civic    1.6  1999    4  manual(m5)   f    28   
104        honda         civic    1.6  1999    4    auto(l4)   f    24   
102        honda         civic    1.6  1999    4  manual(m5)   f    25   
101        honda         civic    1.6  1999    4    auto(l4)   f    24   
190       toyota  camry solara    2.4  2008    4    auto(s5)   f    22   
3           audi            a4    2.0  2008    4  manual(m6)   f    20   
189       toyota  camry solara    2.4  2008    4  manual(m5)   f    21   
194       toyota       corolla    1.8  1999    4    auto(l3)   f    24   
4           audi            a4    2.0  2008    4    auto(av)   f    21   
188       toyota  camry solara    2.2  1999    4  manual(m5)   f    21   
2           audi            a4    1.8  1999    4  manual(m5)   f    21   
1           audi            a4    1.8  1999    4    auto(l5)   f    18   
117      hyundai       tiburon    2.0  1999    4  manual(m5)   f    19   
208   volkswagen           gti    2.0  1999    4  manual(m5)   f    21   
210   volkswagen           gti    2.0  2008    4  manual(m6)   f    21   
211   volkswagen           gti    2.0  2008    4    auto(s6)   f    22   
214   volkswagen         jetta    2.0  1999    4  manual(m5)   f    21   
216   volkswagen         jetta    2.0  2008    4    auto(s6)   f    22   
217   volkswagen         jetta    2.0  2008    4  manual(m6)   f    21   
218   volkswagen         jetta    2.5  2008    5    auto(s6)   f    21   
219   volkswagen         jetta    2.5  2008    5  manual(m5)   f    21   
224   volkswagen    new beetle    2.0  1999    4  manual(m5)   f    21   
142       nissan        altima    2.4  1999    4  manual(m5)   f    21   
227   volkswagen    new beetle    2.5  2008    5    auto(s6)   f    20   
108        honda         civic    2.0  2008    4  manual(m6)   f    21   
103        honda         civic    1.6  1999    4  manual(m5)   f    23   
118      hyundai       tiburon    2.0  2008    4  manual(m5)   f    20   
226   volkswagen    new beetle    2.5  2008    5  manual(m5)   f    20   
10          audi    a4 quattro    2.0  2008    4  manual(m6)   4    20   
193       toyota  camry solara    3.3  2008    6    auto(s5)   f    18   
187       toyota  camry solara    2.2  1999    4    auto(l4)   f    21   
173       subaru   impreza awd    2.5  2008    4  manual(m5)   4    20   
171       subaru   impreza awd    2.5  2008    4    auto(s4)   4    20   
11          audi    a4 quattro    2.0  2008    4    auto(s6)   4    19   
143       nissan        altima    2.4  1999    4    auto(l4)   f    19   
7           audi            a4    3.1  2008    6    auto(av)   f    18   
119      hyundai       tiburon    2.0  2008    4    auto(l4)   f    20   
6           audi            a4    2.8  1999    6  manual(m5)   f    18   
5           audi            a4    2.8  1999    6    auto(l5)   f    16   
215   volkswagen         jetta    2.0  1999    4    auto(l4)   f    19   
225   volkswagen    new beetle    2.0  1999    4    auto(l4)   f    19   
209   volkswagen           gti    2.0  1999    4    auto(l4)   f    19   
91          ford       mustang    3.8  1999    6  manual(m5)   r    18   
93      

j.Create a column named average_mileage that is the mean of the city and highway mileage.

In [266]:
mpg_df['average_mileage'] =  round((mpg_df.highway + mpg_df.city)/2,1)
mpg_df

manufacturer   model  displ  year  cyl       trans drv  city  highway fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17       26  p   

       class  mileage_difference  average_mileage  
1    compact                  11             23.5  
2    compact                   8             25.0  
3    compact                  11             25.5  
4    compact                   9             25.5  
5    compact                  10             21.0  
..       ...                 ...              ...  
230  midsize                   9             23.5  
231  midsize                   8             25.0  
232  midsize                  10             21.0  
233  midsize                   8             22.0  
234  midsize                   9             21.5  

[234 rows x 13 columns]

k.Which dodge car has the best average mileage? The worst?

In [277]:
dodge_car_mpg_df = mpg_df[mpg_df["manufacturer"].str.contains("dodge")]
dodge_car_mpg_df.sort_values("average_mileage")


manufacturer                model  displ  year  cyl       trans drv  city  \
70        dodge  ram 1500 pickup 4wd    4.7  2008    8  manual(m6)   4     9   
66        dodge  ram 1500 pickup 4wd    4.7  2008    8    auto(l5)   4     9   
60        dodge          durango 4wd    4.7  2008    8    auto(l5)   4     9   
55        dodge    dakota pickup 4wd    4.7  2008    8    auto(l5)   4     9   
74        dodge  ram 1500 pickup 4wd    5.9  1999    8    auto(l4)   4    11   
71        dodge  ram 1500 pickup 4wd    5.2  1999    8    auto(l4)   4    11   
64        dodge          durango 4wd    5.9  1999    8    auto(l4)   4    11   
57        dodge    dakota pickup 4wd    5.2  1999    8    auto(l4)   4    11   
72        dodge  ram 1500 pickup 4wd    5.2  1999    8  manual(m5)   4    11   
62        dodge          durango 4wd    5.2  1999    8    auto(l4)   4    11   
69        dodge  ram 1500 pickup 4wd    4.7  2008    8  manual(m6)   4    12   
65        dodge  ram 1500 pickup 4wd    4.7  2008    8  manual(m6)   4    12   
56        dodge    dakota pickup 4wd    5.2  1999    8  manual(m5)   4    11   
44        dodge          caravan 2wd    3.3  2008    6    auto(l4)   f    11   
51        dodge    dakota pickup 4wd    3.9  1999    6    auto(l4)   4    13   
73        dodge  ram 1500 pickup 4wd    5.7  2008    8    auto(l5)   4    13   
68        dodge  ram 1500 pickup 4wd    4.7  2008    8    auto(l5)   4    13   
58        dodge          durango 4wd    3.9  1999    6    auto(l4)   4    13   
59        dodge          durango 4wd    4.7  2008    8    auto(l5)   4    13   
67        dodge  ram 1500 pickup 4wd    4.7  2008    8    auto(l5)   4    13   
61        dodge          durango 4wd    4.7  2008    8    auto(l5)   4    13   
52        dodge    dakota pickup 4wd    3.9  1999    6  manual(m5)   4    14   
63        dodge          durango 4wd    5.7  2008    8    auto(l5)   4    13   
50        dodge    dakota pickup 4wd    3.7  2008    6    auto(l4)   4    14   
53        dodge    dakota pickup 4wd    4.7  2008    8    auto(l5)   4    14   
54        dodge    dakota pickup 4wd    4.7  2008    8    auto(l5)   4    14   
49        dodge    dakota pickup 4wd    3.7  2008    6  manual(m6)   4    15   
46        dodge          caravan 2wd    3.8  1999    6    auto(l4)   f    15   
45        dodge          caravan 2wd    3.8  1999    6    auto(l4)   f    15   
41        dodge          caravan 2wd    3.3  1999    6    auto(l4)   f    16   
40        dodge          caravan 2wd    3.3  1999    6    auto(l4)   f    16   
47        dodge          caravan 2wd    3.8  2008    6    auto(l6)   f    16   
48        dodge          caravan 2wd    4.0  2008    6    auto(l6)   f    16   
43        dodge          caravan 2wd    3.3  2008    6    auto(l4)   f    17   
42        dodge          caravan 2wd    3.3  2008    6    auto(l4)   f    17   
39        dodge          caravan 2wd    3.0  1999    6    auto(l4)   f    17   
38        dodge          caravan 2wd    2.4  1999    4    auto(l3)   f    18   

    highway fl    class  mileage_difference  average_mileage  
70       12  e   pickup                   3             10.5  
66       12  e   pickup                   3             10.5  
60       12  e      suv                   3             10.5  
55       12  e   pickup                   3             10.5  
74       15  r   pickup                   4             13.0  
71       15  r   pickup                   4             13.0  
64       15  r      suv                   4             13.0  
57       15  r   pickup                   4             13.0  
72       16  r   pickup                   5             13.5  
62       16  r      suv                   5             13.5  
69       16  r   pickup                   4             14.0  
65       16  r   pickup                   4             14.0  
56       17  r   pickup                   6             14.0  
44       17  e  minivan                   6             14.0  
51       17  r   

In [280]:
dodge_car_mpg_df.sort_values("average_mileage", ascending= False)

manufacturer                model  displ  year  cyl       trans drv  city  \
38        dodge          caravan 2wd    2.4  1999    4    auto(l3)   f    18   
42        dodge          caravan 2wd    3.3  2008    6    auto(l4)   f    17   
43        dodge          caravan 2wd    3.3  2008    6    auto(l4)   f    17   
39        dodge          caravan 2wd    3.0  1999    6    auto(l4)   f    17   
47        dodge          caravan 2wd    3.8  2008    6    auto(l6)   f    16   
48        dodge          caravan 2wd    4.0  2008    6    auto(l6)   f    16   
40        dodge          caravan 2wd    3.3  1999    6    auto(l4)   f    16   
41        dodge          caravan 2wd    3.3  1999    6    auto(l4)   f    16   
45        dodge          caravan 2wd    3.8  1999    6    auto(l4)   f    15   
46        dodge          caravan 2wd    3.8  1999    6    auto(l4)   f    15   
49        dodge    dakota pickup 4wd    3.7  2008    6  manual(m6)   4    15   
53        dodge    dakota pickup 4wd    4.7  2008    8    auto(l5)   4    14   
54        dodge    dakota pickup 4wd    4.7  2008    8    auto(l5)   4    14   
50        dodge    dakota pickup 4wd    3.7  2008    6    auto(l4)   4    14   
52        dodge    dakota pickup 4wd    3.9  1999    6  manual(m5)   4    14   
63        dodge          durango 4wd    5.7  2008    8    auto(l5)   4    13   
51        dodge    dakota pickup 4wd    3.9  1999    6    auto(l4)   4    13   
73        dodge  ram 1500 pickup 4wd    5.7  2008    8    auto(l5)   4    13   
68        dodge  ram 1500 pickup 4wd    4.7  2008    8    auto(l5)   4    13   
58        dodge          durango 4wd    3.9  1999    6    auto(l4)   4    13   
59        dodge          durango 4wd    4.7  2008    8    auto(l5)   4    13   
67        dodge  ram 1500 pickup 4wd    4.7  2008    8    auto(l5)   4    13   
61        dodge          durango 4wd    4.7  2008    8    auto(l5)   4    13   
69        dodge  ram 1500 pickup 4wd    4.7  2008    8  manual(m6)   4    12   
65        dodge  ram 1500 pickup 4wd    4.7  2008    8  manual(m6)   4    12   
56        dodge    dakota pickup 4wd    5.2  1999    8  manual(m5)   4    11   
44        dodge          caravan 2wd    3.3  2008    6    auto(l4)   f    11   
62        dodge          durango 4wd    5.2  1999    8    auto(l4)   4    11   
72        dodge  ram 1500 pickup 4wd    5.2  1999    8  manual(m5)   4    11   
64        dodge          durango 4wd    5.9  1999    8    auto(l4)   4    11   
57        dodge    dakota pickup 4wd    5.2  1999    8    auto(l4)   4    11   
71        dodge  ram 1500 pickup 4wd    5.2  1999    8    auto(l4)   4    11   
74        dodge  ram 1500 pickup 4wd    5.9  1999    8    auto(l4)   4    11   
66        dodge  ram 1500 pickup 4wd    4.7  2008    8    auto(l5)   4     9   
60        dodge          durango 4wd    4.7  2008    8    auto(l5)   4     9   
55        dodge    dakota pickup 4wd    4.7  2008    8    auto(l5)   4     9   
70        dodge  ram 1500 pickup 4wd    4.7  2008    8  manual(m6)   4     9   

    highway fl    class  mileage_difference  average_mileage  
38       24  r  minivan                   6             21.0  
42       24  r  minivan                   7             20.5  
43       24  r  minivan                   7             20.5  
39       24  r  minivan                   7             20.5  
47       23  r  minivan                   7             19.5  
48       23  r  minivan                   7             19.5  
40       22  r  minivan                   6             19.0  
41       22  r  minivan                   6             19.0  
45       22  r  minivan                   7             18.5  
46       21  r  minivan                   6             18.0  
49       19  r   pickup                   4             17.0  
53       19  r   pickup                   5             16.5  
54       19  r   pickup                   5             16.5  
50       18  r   pickup                   4             16.0  
52       17  r   

a.How many rows and columns are there?

    b.What are the data types of each column?

    c.Summarize the dataframe with .info and .describe

    d.Rename the cty column to city.

    e.Rename the hwy column to highway.

    f.Do any cars have better city mileage than highway mileage?

    g.Create a column named mileage_difference this column should contain the difference between highway and city mileage for each car.

    h.Which car (or cars) has the highest mileage difference?

    i.Which compact class car has the lowest highway mileage? The best?

    j.Create a column named average_mileage that is the mean of the city and highway mileage.

    k.Which dodge car has the best average mileage? The worst?
